alg/teca_normalize_coordinates.cxx

```diff
@@ -0,0 +1,440 @@
+#include "teca_normalize_coordinates.h"
+
+#include "teca_cartesian_mesh.h"
+#include "teca_array_collection.h"
+#include "teca_variant_array.h"
+#include "teca_metadata.h"
+#include "teca_coordinate_util.h"
+
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <set>
+#include <cmath>
+
+#if defined(TECA_HAS_BOOST)
+#include <boost/program_options.hpp>
+#endif
+
+using std::cerr;
+using std::endl;
+
+//#define TECA_DEBUG
+
+struct teca_normalize_coordinates::internals_t
+{
+    internals_t() {}
+    ~internals_t() {}
+
+    static
+    p_teca_variant_array normalize_axis(const const_p_teca_variant_array &x);
+
+    static
+    void normalize_extent(p_teca_variant_array out_x,
+        p_teca_variant_array out_y, p_teca_variant_array out_z,
+        unsigned long *whole_extent, unsigned long *extent_in,
+        unsigned long *extent_out);
+
+    static
+    void normalize_variables(bool normalize_x,
+        bool normalize_y, bool normalize_z, unsigned long *extent,
+        p_teca_array_collection data);
+};
+
+
+// --------------------------------------------------------------------------
+p_teca_variant_array teca_normalize_coordinates::internals_t::normalize_axis(
+    const const_p_teca_variant_array &x)
+{
+    unsigned long nx = x->size();
+    unsigned long x1 = nx - 1;
+
+    NESTED_TEMPLATE_DISPATCH(const teca_variant_array_impl,
+        x.get(), _C,
+
+        const NT_C *px = dynamic_cast<TT_C*>(x.get())->get();
```

Comment:
Lines 55-66 are a sample block of code where it would be useful to add a bit of commentary indicating the function of the code block.  On reading through the code, I think that you are checking whether the coordinate needs to be reversed to be ordered negative->positive, but I'm not sure.